src/json.c

```diff
@@ -0,0 +1,1150 @@
+//
+//  Copyright (C) 2011-2018  Nick Gasson
+//  Copyright (C) 2019       Sebastien Van Cauwenberghe
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//  This program is distributed in the hope that it will be useful,
+//  but WITHOUT ANY WARRANTY; without even the implied warranty of
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU General Public License for more details.
+//
+//  You should have received a copy of the GNU General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+#include "phase.h"
+#include "util.h"
+#include "hash.h"
+#include "common.h"
+#include "rt/netdb.h"
+#include "json.h"
+
+#include <assert.h>
+#include <string.h>
+#include <stdio.h>
+#include <ctype.h>
+#include <inttypes.h>
+#include <stdlib.h>
+
+LCOV_EXCL_START
+
+static JsonNode *dump_expr(tree_t t);
+static JsonNode *dump_stmt(tree_t t);
+static JsonNode *dump_port(tree_t t);
+static JsonNode *dump_decl(tree_t t);
+static JsonNode *dump_decls(tree_t t);
+
+typedef tree_t (*get_fn_t)(tree_t, unsigned);
+
+JsonNode *base_node = NULL;
+
+static void cannot_dump(tree_t t, const char *hint)
+{
+   printf("\n");
+   fflush(stdout);
+   fatal("cannot dump %s kind %s", hint, tree_kind_str(tree_kind(t)));
+}
+
+static void add_lineno(JsonNode *obj, tree_t node) {
+   const loc_t *loc = tree_loc(node);
+   json_append_member(obj, "ln", json_mknumber(loc->first_line));
+}
+
+static void add_filename(JsonNode *obj, tree_t node) {
```

Comment:
I suggest you make one function `add_loc()` that stores in the filename as well as the line number to avoid any confusion

Reply:
Thanks for the suggestion, I wanted to add the filename only on entities, architectures, packages and packages bodies. 
I used the assumptions that the other nodes below would only need the linenumber as the file is the same.